source/fab/steps/cleanup_prebuilds.py

```diff
@@ -0,0 +1,179 @@
+
+        else:
+            # Make sure the file system can give us access times - overkill?
+            if not check_fs_access_time(config.prebuild_folder):
+                logger.error("file access time not available, aborting housekeeping")
+                return
+
+            # get the file access time for every artefact
+            prebuilds_ts = \
+                dict(zip(prebuild_files, self.run_mp(prebuild_files, get_access_time)))  # type: ignore
+
+            # work out what to delete
+            to_delete = self.by_age(prebuilds_ts, current_files=artefact_store[CURRENT_PREBUILDS])
+            to_delete |= self.by_version_age(prebuilds_ts, current_files=artefact_store[CURRENT_PREBUILDS])
+
+            # delete them all
+            self.run_mp(to_delete, os.remove)
+            num_removed = len(to_delete)
+
+        logger.info(f'removed {num_removed} prebuild files')
+
+    def by_age(self, prebuilds_ts: Dict[Path, datetime], current_files: Iterable[Path]) -> Set[Path]:
+        to_delete = set()
+
+        if self.older_than:
+            most_recent_ts = max(prebuilds_ts.values())
+            oldest_ts_allowed = most_recent_ts - self.older_than
+
+            for f, ts in prebuilds_ts.items():
+                if ts < oldest_ts_allowed:
+                    # don't delete if it's still current
+                    if f in current_files:
+                        logger.debug(f"old file is still current {f}")
+                        continue
+                    logger.info(f"old file {f}")
+                    to_delete.add(f)
+
+        return to_delete
+
+    def by_version_age(self, prebuilds_ts: Dict[Path, datetime], current_files: Iterable[Path]) -> Set[Path]:
+        to_delete = set()
+
+        if self.n_versions:
+            # group prebuild files by originating artefact, <stem>.*.<suffix>
+            pb_file_groups = get_prebuild_file_groups(prebuilds_ts.keys())
+
+            # delete the n oldest in each group
+            for pb_group in pb_file_groups.values():
+                by_age = sorted(pb_group, key=lambda f: prebuilds_ts[f], reverse=True)
+                for f in by_age[self.n_versions:]:
+                    # don't delete if it's still current
+                    if f in current_files:
+                        logger.debug(f"old version is still current {f}")
+                        continue
+                    logger.debug(f"old version {f}")
+                    to_delete.add(f)
+
+        return to_delete
+
+
+def remove_all_unused(found_files: Iterable[Path], current_files: Iterable[Path]):
+    num_removed = 0
+
+    for f in found_files:
+        if f not in current_files:
+            logger.info(f"unused {f}")
+            os.remove(f)
+            num_removed += 1
+
+    return num_removed
+
+
+# this whole function might be overkill...how likely is this, and do we care?
+def check_fs_access_time(folder=None) -> bool:
```

Comment:
I think this check is a good safety net but as you observe there's quite a lot of it. It will almost certainly never be called so the real question is, how to test it? If it can't be effectively tested then we are carrying technical debt around, if it is well tested then its size is not so much of an issue.

Reply:
I don't know how to test this properly without something like [TestingBot](https://testingbot.com/), which would be woahverkill. Instead, we could just document the need for a file system which reports access time and drop this. I'm leaning that way now. I think you are too so I'll remove this feature.